client/src/modules/cash/dashboard/dashboard.chart.js

```diff
@@ -0,0 +1,309 @@
+
+    // clear canvas for previous chart rendering
+    if (vm.monthChart) { vm.monthChart.destroy(); }
+
+    vm.monthChart = new Chart(ctx, {
+      type : 'bar',
+      options : chartOptions($translate.instant('DASHBOARD.CASH.CHART_MONTH')),
+      data : {
+        labels : newData.dataset.labels,
+        datasets : [
+          {
+            label : $translate.instant('DASHBOARD.CASH.PAYMENT'),
+            data : newData.dataset.data,
+            backgroundColor : 'rgba(54, 162, 235, 0.2)',
+            borderColor : 'rgba(54, 162, 235, 1)',
+            borderWidth : 1,
+          },
+
+          {
+            label : $translate.instant('DASHBOARD.CASH.REVERSED'),
+            data : newData.reversedDataset.data,
+            backgroundColor : 'rgba(255, 99, 132, 0.2)',
+            borderColor : 'rgba(255,99,132,1)',
+            borderWidth : 1,
+          },
+        ],
+      },
+    });
+  }
+
+  /**
+   * Generate Group chart
+   * @param {object} data
+   */
+  function generateGroupChart(data) {
+    var dataset = getCountData(data, 'group', 'name');
+    var ctx = document.getElementById('groupChart').getContext('2d');
+
+    // clear canvas for previous chart rendering
+    if (vm.groupChart) { vm.groupChart.destroy(); }
+
+    vm.groupChart = new Chart(ctx, {
+      type : 'pie',
+      options : chartOptions($translate.instant('DASHBOARD.CASH.DEBTOR_GROUP')),
+      data : {
+        labels : dataset.labels,
+        datasets : [
+          {
+            label : $translate.instant('DASHBOARD.CASH.PAYMENT'),
+            data : dataset.data,
+            backgroundColor : [
+              '#60c6e0', '#0b97b5', '#287181', '#22c36a', '#207647',
+              '#db9b2b', '#634614', '#e91945', '#be1cb4', '#e3e934',
+            ],
+          },
+        ],
+      },
+    });
+  }
+
+  /** chart options */
+  function chartOptions(text) {
+    return {
+      title : {
+        display : true,
+        text : text,
+        fontSize : 20,
+      },
+      legend : {
+        display : true,
+      },
+    };
+  }
+
+  /** get hour data */
+  function getHourData(dataset, field) {
+    var out = { labels : [], data : [] };
+    if (!dataset[field]) { return out; }
+
+    dataset[field].forEach(function (item) {
+      out.labels.push(String(item.hour).concat('H'));
+      out.data.push(item.count);
+    });
+
+    return out;
+  }
+
+  /** get counted data */
+  function getCountData(dataset, field, label) {
+    var out = { labels : [], data : [] };
+    if (!dataset[field]) { return out; }
+
+    dataset[field].forEach(function (item) {
+      out.labels.push(item[label]);
+      out.data.push(item.count);
+    });
+
+    return out;
+  }
+
+  /**
+   * @function uniformize data
+   * @param {object} data - { data: [], labels: [] }
+   * @param {object} reversedData - { data: [], labels: [] }
+   * @param {string} periodGroup - string in ('hour', 'day', 'month')
+   */
+  function uniformize(data, reversedData, periodGroup) {
```

Comment:
Putting this information in a function is the right idea. :+1:

However, this function is pretty complex.  We will likely also need it on multiple dashboards.

I propose you move this functionality into a service ... maybe `DashboardDataService`.  The `hours`, `dayofweek`, `months`, etc, can be private variables.  You can then expose methods to sort number arrays, make uniform datasets, and convert an array to unique values.